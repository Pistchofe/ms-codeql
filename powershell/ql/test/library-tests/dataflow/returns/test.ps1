function callSourceOnce {
    Source "1"
}

$x = callSourceOnce
Sink $x # $ MISSING: hasValueFlow=1

function callSourceTwice {
    Source "2"
    Source "3"
}

$x = callSourceTwice
Sink $x # $ MISSING: hasValueFlow=2 hasValueFlow=3

function returnSource1 {
    return Source "4"
}

$x = returnSource1
Sink $x # $ MISSING: hasValueFlow=4

function returnSource2 {
    $x = Source "5"
    $x
    $y = Source "6"
    return $y
}

$x = returnSource2
Sink $x # $ MISSING: hasValueFlow=5 hasValueFlow=6