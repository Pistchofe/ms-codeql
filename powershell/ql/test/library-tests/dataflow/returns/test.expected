models
edges
| test.ps1:2:5:2:14 | Call to Source | test.ps1:5:6:5:19 | Call to callSourceOnce | provenance |  |
| test.ps1:5:6:5:19 | Call to callSourceOnce | test.ps1:6:6:6:7 | x | provenance |  |
| test.ps1:9:5:9:14 | Call to Source | test.ps1:13:6:13:20 | Call to callSourceTwice [element] | provenance |  |
| test.ps1:10:5:10:14 | Call to Source | test.ps1:13:6:13:20 | Call to callSourceTwice [element] | provenance |  |
| test.ps1:13:6:13:20 | Call to callSourceTwice [element] | test.ps1:15:6:15:7 | x [element] | provenance |  |
| test.ps1:13:6:13:20 | Call to callSourceTwice [element] | test.ps1:16:6:16:7 | x [element] | provenance |  |
| test.ps1:15:6:15:7 | x [element] | test.ps1:15:6:15:10 | ...[...] | provenance |  |
| test.ps1:16:6:16:7 | x [element] | test.ps1:16:6:16:10 | ...[...] | provenance |  |
| test.ps1:38:9:38:18 | Call to Source | test.ps1:42:6:42:21 | Call to callSourceInLoop [element] | provenance |  |
| test.ps1:42:6:42:21 | Call to callSourceInLoop [element] | test.ps1:43:6:43:7 | x [element] | provenance |  |
| test.ps1:42:6:42:21 | Call to callSourceInLoop [element] | test.ps1:44:6:44:7 | x [element] | provenance |  |
| test.ps1:43:6:43:7 | x [element] | test.ps1:43:6:43:10 | ...[...] | provenance |  |
| test.ps1:44:6:44:7 | x [element] | test.ps1:44:6:44:10 | ...[...] | provenance |  |
nodes
| test.ps1:2:5:2:14 | Call to Source | semmle.label | Call to Source |
| test.ps1:5:6:5:19 | Call to callSourceOnce | semmle.label | Call to callSourceOnce |
| test.ps1:6:6:6:7 | x | semmle.label | x |
| test.ps1:9:5:9:14 | Call to Source | semmle.label | Call to Source |
| test.ps1:10:5:10:14 | Call to Source | semmle.label | Call to Source |
| test.ps1:13:6:13:20 | Call to callSourceTwice [element] | semmle.label | Call to callSourceTwice [element] |
| test.ps1:15:6:15:7 | x [element] | semmle.label | x [element] |
| test.ps1:15:6:15:10 | ...[...] | semmle.label | ...[...] |
| test.ps1:16:6:16:7 | x [element] | semmle.label | x [element] |
| test.ps1:16:6:16:10 | ...[...] | semmle.label | ...[...] |
| test.ps1:38:9:38:18 | Call to Source | semmle.label | Call to Source |
| test.ps1:42:6:42:21 | Call to callSourceInLoop [element] | semmle.label | Call to callSourceInLoop [element] |
| test.ps1:43:6:43:7 | x [element] | semmle.label | x [element] |
| test.ps1:43:6:43:10 | ...[...] | semmle.label | ...[...] |
| test.ps1:44:6:44:7 | x [element] | semmle.label | x [element] |
| test.ps1:44:6:44:10 | ...[...] | semmle.label | ...[...] |
subpaths
testFailures
| test.ps1:23:9:23:26 | # $ hasValueFlow=4 | Missing result: hasValueFlow=4 |
| test.ps1:33:12:33:54 | # $ hasValueFlow=5 SPURIOUS: hasValueFlow=6 | Fixed spurious result: hasValueFlow=6 |
| test.ps1:33:12:33:54 | # $ hasValueFlow=5 SPURIOUS: hasValueFlow=6 | Missing result: hasValueFlow=5 |
| test.ps1:34:12:34:54 | # $ hasValueFlow=6 SPURIOUS: hasValueFlow=5 | Fixed spurious result: hasValueFlow=5 |
| test.ps1:34:12:34:54 | # $ hasValueFlow=6 SPURIOUS: hasValueFlow=5 | Missing result: hasValueFlow=6 |
#select
| test.ps1:6:6:6:7 | x | test.ps1:2:5:2:14 | Call to Source | test.ps1:6:6:6:7 | x | $@ | test.ps1:2:5:2:14 | Call to Source | Call to Source |
| test.ps1:15:6:15:10 | ...[...] | test.ps1:9:5:9:14 | Call to Source | test.ps1:15:6:15:10 | ...[...] | $@ | test.ps1:9:5:9:14 | Call to Source | Call to Source |
| test.ps1:15:6:15:10 | ...[...] | test.ps1:10:5:10:14 | Call to Source | test.ps1:15:6:15:10 | ...[...] | $@ | test.ps1:10:5:10:14 | Call to Source | Call to Source |
| test.ps1:16:6:16:10 | ...[...] | test.ps1:9:5:9:14 | Call to Source | test.ps1:16:6:16:10 | ...[...] | $@ | test.ps1:9:5:9:14 | Call to Source | Call to Source |
| test.ps1:16:6:16:10 | ...[...] | test.ps1:10:5:10:14 | Call to Source | test.ps1:16:6:16:10 | ...[...] | $@ | test.ps1:10:5:10:14 | Call to Source | Call to Source |
| test.ps1:43:6:43:10 | ...[...] | test.ps1:38:9:38:18 | Call to Source | test.ps1:43:6:43:10 | ...[...] | $@ | test.ps1:38:9:38:18 | Call to Source | Call to Source |
| test.ps1:44:6:44:10 | ...[...] | test.ps1:38:9:38:18 | Call to Source | test.ps1:44:6:44:10 | ...[...] | $@ | test.ps1:38:9:38:18 | Call to Source | Call to Source |
